device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java

```diff
@@ -1142,23 +1297,48 @@ private void updateStatus(IotHubConnectionStatus newConnectionStatus, IotHubConn
             //invoke connection status callbacks
             this.log.debug("Invoking connection status callbacks with new status details");
             invokeConnectionStateCallback(newConnectionStatus, reason);
-            invokeConnectionStatusChangeCallback(newConnectionStatus, reason, throwable);
+
+            if (deviceClientConfigs.size() < 2 || newConnectionStatus != IotHubConnectionStatus.CONNECTED)
+            {
+                // When multiplexing, a different method will notify each device-specific callback when that device is online,
+                // but in cases when the tcp connection is lost and everything is disconnected retrying or disconnected, this is where the
+                // callback should be fired
+                invokeConnectionStatusChangeCallback(newConnectionStatus, reason, throwable);
+
+                for (DeviceClientConfig config : deviceClientConfigs)
+                {
+                    deviceConnectionStates.put(config.getDeviceId(), newConnectionStatus);
+                }
+            }
+
             this.deviceIOConnectionStatusChangeCallback.execute(newConnectionStatus, reason, throwable, null);
+        }
+    }
 
-            if (newConnectionStatus == IotHubConnectionStatus.CONNECTED)
+    private void updateStatus(IotHubConnectionStatus newConnectionStatus, IotHubConnectionStatusChangeReason reason, Throwable throwable, String deviceId)
+    {
+        if (this.deviceConnectionStates.containsKey(deviceId) && this.deviceConnectionStates.get(deviceId) != newConnectionStatus)
+        {
+            if (throwable == null)
+            {
+                this.log.debug("Updating device {} status to new status {} with reason {}", deviceId, newConnectionStatus, reason);
+            }
+            else
             {
-                //Codes_SRS_IOTHUBTRANSPORT_28_007: [This function shall reset currentReconnectionAttempt and reconnectionAttemptStartTimeMillis if connection status is changed to CONNECTED.]
-                this.currentReconnectionAttempt = 0;
-                this.reconnectionAttemptStartTimeMillis = 0;
+                this.log.warn("Updating device {} status to new status {} with reason {}", deviceId, newConnectionStatus, reason, throwable);
+            }
+
+            synchronized (this.multiplexingDeviceStateLock)
+            {
+                this.deviceConnectionStates.put(deviceId, newConnectionStatus);
+
+                this.log.debug("Invoking connection status callbacks with new status details");
```

Comment:
Do we not want to log each device's connection status?

Reply:
We do log them above